internal/blobtesting/versionedmap.go

```diff
@@ -0,0 +1,248 @@
+package blobtesting
+
+import (
+	"bytes"
+	"context"
+	"sort"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/kopia/kopia/internal/clock"
+	"github.com/kopia/kopia/repo/blob"
+)
+
+type entry struct {
+	value          []byte
+	mtime          time.Time
+	retentionTime  time.Time
+	isDeleteMarker bool
+}
+
+type versionedEntries map[blob.ID][]*entry
+
+type versionedMapStorage struct {
+	data    versionedEntries
+	timeNow func() time.Time
+	mutex   sync.RWMutex
+}
+
+func (s *versionedMapStorage) getLatestByID(id blob.ID, writeIntent bool) (*entry, error) {
+	versions, ok := s.data[id]
+	if !ok {
+		return nil, blob.ErrBlobNotFound
+	}
+
+	// get the latest version and if it is a delete marker then simulate
+	// not-found
+	e := versions[len(versions)-1]
+	if e.isDeleteMarker {
+		return nil, blob.ErrBlobNotFound
+	}
+
+	if writeIntent && !e.retentionTime.IsZero() && e.retentionTime.After(s.timeNow()) {
+		return nil, errors.New("cannot alter object before retention period expires")
+	}
```

Comment:
I'm not sure I understand this part. This seems to be related to object locking.
Can this be done in the caller and then remove the `writeIntent` parameter?

The operation that should fail when the retention has not elapsed is deleting a blob version. `PutBlob` and `DeleteBlob` should succeed even if the retention has not elapsed.

Reply:
> Can this be done in the caller and then remove the writeIntent parameter?

Yes. The `writeIntent` allowed reuse of the same check. If `DeleteBlob` should succeed then we can move this to `PutBlob` because it is not a common check elsewhere.